flex_model/core/wrapper.py

```diff
@@ -213,42 +372,89 @@ def __init__(
+                if handle is not None and hook_fn in hooks:
+                    handle.remove()
+                    self._module_to_hook_fns_map[module][hook_fn] = None
 
-    @_make_group.register
-    def _(self, group_constructor: str):
-        # TODO: Pattern match module names for already-registered hook functions.
-        raise NotImplementedError
+    def _flush_pipeline(self) -> None:
+        """Gather tensors from all pipeline stages to the first stage."""
+        if (
+            torch.distributed.is_initialized()
+            and dist.in_pipeline_parallel_group()
+            and dist.get_activation_pipeline_parallel_world_size() > 1
+        ):
+            gathered_acts = dist.gather_pipeline_parallel_tensor_dicts(self.output_ptr)
 
-    @_make_group.register
-    def _(self, group_constructor: list, group_alias: str):
-        # TODO: Construct group explicitly using hook functions.
-        raise NotImplementedError
+            # Rank 0 accumulates the activation tensors.
+            if dist.get_activation_pipeline_parallel_rank() == 0:
+                self.output_ptr.update(gathered_acts)
+
+            # Other ranks reset their collections for the next forward pass.
+            else:
+                self.output_ptr = {}
+
+    def forward(
+        self,
+        *args,
+        groups: Union[str, List[str]] = "all",
+        complement: bool = False,
+        **kwargs,
+    ) -> Any:
+        """Run a forward pass of the model with hooks enabled by default."""
+        active, inactive = self._hook_fn_group_manager.bisect(groups)
+
+        if not complement:
+            self._enable_hooks(active)
+            self._disable_hooks(inactive)
+        else:
+            self._enable_hooks(inactive)
+            self._disable_hooks(active)
+
+        outputs = self.module(*args, **kwargs)
+
+        # Post-forward cleanup.
+        self._flush_pipeline()
+
+        return outputs
 
     def _register_hook_impl(self, hook_function: HookFunction) -> None:
-        """Registers hook to underlying pytorch module."""
+        """Pytorch `nn.Module` hook function registration implementation."""
+        # TODO: `get_module` should work with the parameter tensors in the case
+        #       of `register_hook(Tensor)`. Needs verification.
         module = _get_module(self.module, hook_function.module_name)
 
         # Register hook function to pt module.
         register_fn = getattr(
             module, self._hook_type_to_pt_attr[hook_function._hook_type], None
         )
         handle = register_fn(hook_function)
-        self.hook_functions["all"][module][hook_function] = handle
 
-    def _hook_registration_prologue(
+        # All registered hooks are members of the "all" hook function group.
+        self._module_to_hook_fns_map[module][hook_function] = handle
+        self._hook_fn_group_manager.update(hook_function, "all")
+
+    def _register_hook_prologue(
```

Comment:
I think John commented on this, too, but is there a better name than prologue?

Reply:
Would argue that prologue is an exact definition, since it does all the setup and checks before running the actual hook registration to the module/tensor.